### PR TITLE
Describe problems for functor

### PR DESCRIPTION
**Pull request by reyoung**
Part one of `Functor Design Doc`.  It describes what problems should be resolved.

Related issue #3894

### CODE REVIEW

File: doc/design/functor.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Design Doc: Operator Functor
+
```
Comment (wangkuiyi) [inline doc/design/functor.md]: Let's use an example to explain the problem:

```c++


template <typename T>
T tanh(T x) {
  return ...;
}

template <typename T>
T sigmoid(T x) {
  return (tanh(x) + 1) / 2;
}

template <typename Place>
class SigmodOpKernel : public framework::OpKernel {};

template <>
class SigmodOpKernel<CPUPlace> : public framework::OpKernel {};
 public:
  void Compute(ctx) {
    std::transform(
        ctx.Input<Tensor>("X")->data(),
        ctx.Input<Tensor>("X")->data() + product(ctx.Input<Tensor>("X")->dims()),
        ctx.Output<Tensor>("Out")->mutable_data(),
        sigmoid);
  }
};

template <>
class SigmodOpKernel<GPUPlace> : public framework::OpKernel {};
 public:
  void Compute(ctx) {
    thrust::transform(
        ctx.Input<Tensor>("X")->data(),
        ctx.Input<Tensor>("X")->data() + product(ctx.Input<Tensor>("X")->dims()),
        ctx.Output<Tensor>("Out")->mutable_data(),
        sigmoid);
  }
};
```

File: doc/design/functor.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Design Doc: Operator Functor
```
Comment (wangkuiyi) [inline doc/design/functor.md]: > # Design Doc: Functions

> Following the overall [design](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/functions_operators_layers.md), this document is about the functions and how they compose into operators.

> Basically, to the ease of definitions, functions operate on scalar values. But most operators operate on Tensors. For example:

**Comment by QiJune** [comment]
Dose this doc only focus on element-wise operators or try to propose a general solution?
 std::transform is not suitable for all operations.

**Comment by reyoung** [comment]
Closed since #3901

**Pull request closed without merging**